[REVIEW] Adding new Dask CUDA serializers to Naive Bayes (and another trivial perf update)

cjnolet opened the pull request.

JohnZed commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -242,25 +242,24 @@ def predict(self, X):
-                                                broadcast=True,
-                                                workers=[w]))
-                       for w, p in x_worker_parts.items()])
+        models = self.client_.scatter(self.model_,
```
does this need the traverse=True treatment?

cjnolet replied:
I don’t believe so. Scatter doesn’t have a traverse option and the tests are not failing from this code path. It looks like it’s on dask.delayed, client.submit, and client.compute. Hoping @jakirkham might advise.

jakirkham commented:
So some of these changes will require the very latest versions of Dask + Distributed (planned to be released tomorrow). Is this something we are ok with? Happy to review either way, but my advice will be different depending on that answer 🙂

JohnZed commented:
Per @jakirkham 's comment, we should strive to be compatible with slightly-older dask versions _unless_ we get buy in from ops right away that other libs are requiring this version too... I don't want to intro a major library dependency change on burndown day unless it's already planned.

kkraus14 commented:
> Per @jakirkham 's comment, we should strive to be compatible with slightly-older dask versions _unless_ we get buy in from ops right away that other libs are requiring this version too... I don't want to intro a major library dependency change on burndown day unless it's already planned.

@JohnZed from the cuDF side we're super dependent on the latest Dask / Distributed anyway so I think we're going to have to bump this dependency in cuDF regardless.

cjnolet commented:
@JohnZed @jakirkham are we good to get these changes in for 0.13?

jakirkham commented:
I have some draft review comments. Was just waiting to hear what we were doing with the Dask version. @JohnZed, are we bumping the Dask version or trying to maintain backwards compatibility?

cjnolet commented:
rerun tests

quasiben commented:
@JohnZed @kkraus14 2.12 was released an hour ago if we are going to make a hard requirement in cuDF we should probably get that done soon and alert OPs to an upcoming change as we are entering burndown

jakirkham commented:
One high-level comment I have would be to move all of the `register_generic` calls to [`cuml.comm.serialize`]( https://github.com/rapidsai/cuml/blob/4a7070f123500459bafea52f069339b8a560b4d2/python/cuml/comm/serialize.py#L56 ). Would also move the `copyreg` bit with `MatDescriptor` there too. The benefit is that Dask will pick these up as [part of startup]( https://github.com/dask/distributed/blob/15591929f3a6b7b390a7a5394e1f53fe6a6c16f4/distributed/protocol/__init__.py#L119 ). Then we don't need to go through the effort of doing this on the client and each worker ourselves. Would imagine this will simplify the code here a fair bit too.

cjnolet commented:
@jakirkham, 

Definitely a great idea to consolidate the MultinomialNB in `cuml.comm.serialize` 

I'm not sure the `MatDescriptor` workaround belongs there, though, since it seems the contents of `cuml.comm.serialize` only get executed when an object in the `cuml` base package gets serialized. 

I installed the latest Dask & Distributed from source and noticed the `MatDescriptor` workaround is already implemented there. I've verified that the version in `distributed` is actually being registered, however, I seem to be encountering behavior similar to #1734 (even though I'm using CUDA 10.0). Registering the serializer for the mat descriptor on the cuML side appears to make this error go away.

jakirkham commented:
No objection to keeping the `MatDescriptor` pickling workaround here. Thoughts on registering it in `cuml.comm.serialize` in addition to where you are registering it now?

jakirkham commented on python/cuml/test/dask/test_utils.py:
```diff
@@ -0,0 +1,52 @@
+
+    stype, sbytes = serialize(mnb, serializers=['dask'])
+
+    assert stype['serializer'] == 'dask'
```
Not sure if we want to test `pickle` too, but this should do that for us if we want.

```suggestion
    assert stype['serializer'] == 'dask'

    stype, sbytes = serialize(mnb, serializers=['pickle'])

    assert stype['serializer'] == 'pickle'
```

jakirkham commented on python/cuml/comm/serialize.py:
```diff
@@ -30,6 +32,17 @@
+    register_generic(MultinomialNB, "cuda",
+                     cuda_serialize, cuda_deserialize)
+    register_generic(MultinomialNB, "dask",
+                     dask_serialize, dask_deserialize)
```
For backwards compatibility with older versions of Dask, would suggest the following change.

```suggestion
    try:
        # Requires Dask Distributed 2.12.0+
        register_generic(MultinomialNB, "cuda",
                         cuda_serialize, cuda_deserialize)
    except TypeError:
        pass
    register_generic(MultinomialNB)
```

cjnolet commented:
@jakirkham, thank you for the latest round of review. I've moved the MatDescriptor patch logic into `cuml.comm.serialize` and I'm now explicitly importing `cuml.comm.serialize` in Dask Naive Bayes and one of our utility classes. Also added a test for the `MatDescriptor` workaround and relevant comments. 

When you get a moment, mind taking another look?

jakirkham commented on python/cuml/comm/serialize.py:
```diff
@@ -25,11 +27,34 @@
+    except TypeError:
+        pass
+
+    register_generic(MultinomialNB)
```
So this is the same as `register_generic(MultinomialNB, "dask", dask_serialize, dask_deserialize)`. If we prefer to keep that line above, maybe we can move this into the `except TypeError` block?

```suggestion
        register_generic(MultinomialNB)
```

jakirkham commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -217,7 +217,10 @@ def _get_part(parts, idx):
+    def predict(self, X):
+        # TODO: Once cupy sparse arrays are fully supported
+        #  underneath Dask arrays,
+        # this can extend DelayedPredictionMixin.
```
Can you please provide context as to what is still needed? Any open issues or PRs? Admittedly these likely won't get solved near term, but it would be good to keep track of all of them somewhere?

jakirkham commented on python/cuml/dask/common/dask_arr_utils.py:
```diff
@@ -164,24 +164,29 @@ def to_sp_dask_array(cudf_or_array, client=None):
+        # serializer, which doesn't know about cupy sparse arrays. This needs
+        # to be updated on the distributed side. Once that is complete, we
+        # can use `client.scatter` here.
         cudf_or_array = client.submit(_x_p, cudf_or_array)
```
Do we have an issue somewhere tracking this?

jakirkham commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -232,7 +235,48 @@ def predict(self, X, delayed=True, parallelism=5):
+                                      broadcast=True,
+                                      direct=True,
+                                      hash=False,
+                                      workers=list(x_worker_parts.keys()))
```
Will `scatter` also have issues with models if it has issues with CuPy sparse matrices?

cjnolet replied:
No. In fact, the Naive Bayes implementation doesn't actually require serialization of any CuPy sparse arrays and the model itself doesn't store any sparse arrays as state. We need the CuPy sparse array serialization capability so that users can build and manipulate Dask arrays which are backed by CuPy sparse arrays. Currently this is not straightforward, mostly because of a lack of CuPy functionality, but the capability for CuPy sparse arrays to be serializable within the Dask environment is also necessary.

jakirkham commented:
Generally LGTM. One last suggested change above. Otherwise just a few lingering questions (mostly about issues for future follow-up).

cjnolet commented:
@dantegd, @JohnZed any chance I could get your feedback and/or approval on this?

jakirkham commented:
LGTM. Thanks for working on this @cjnolet! 😄

JohnZed commented:
Discussing with @quasiben and @kkraus14 here and building on Keith's comment above... since other libs for 0.13 will require Dask 2.12+, we _can_ depend on a later version of Dask. But should update the conda recipe as well.

jakirkham commented:
FWIW @JohnZed the changes in this PR no longer require Dask 2.12+, but will be more performant if it is available.

jakirkham commented on python/cuml/comm/serialize.py:
```diff
@@ -25,11 +27,37 @@
+
+    # Registers 'dask' serialization by default
+    register_generic(MultinomialNB)
+    register_generic(cuml.Base)
```
Was thinking of suggesting this, but didn't know if you had a reason not to do this. If you make this change, can drop 
`register_generic` calls on `MultinomialNB` if it is a subclass of `cuml.Base`. This goes both for this line and the one in the `try` section above.

cjnolet replied:
Unfortunatley, MultinomialNB doesn't extend `cuml.Base` so the separate registration is needed at this time. At one point I had a comment in here about this but between other refactors, merges, and moving stuff around in this PR, it was lost. 

I'll open up an issue and reference it here.

jakirkham replied:
Ah ok. Thanks for the context.

Do the other models contain attributes that we can already serialize or are there other things we are still missing?

cjnolet replied:
Naive Bayes is a little different than a lot of the other models because it 1) is 100% Python on the cuML side, and 2) accepts sparse arrays as input. Like some of the other models, however, it has not yet been refactored to use the new cuml array. This refactor will hopefully be conquered in 0.14, along with many others: https://github.com/rapidsai/cuml/issues/1834

JohnZed commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -232,7 +236,48 @@ def predict(self, X, delayed=True, parallelism=5):
+
+        key = uuid1()
+        futures = [(wf[0],
+                    self.client_.submit(MultinomialNB._get_size,
```
Could DistributedDataHandler.calculate_parts_to_sizes handle this?

cjnolet replied:
It can, and I originally had refactored this to use that, however I was getting segfaults during benchmarks from the use of sparse arrays. I reverted the changes for now, but have created https://github.com/rapidsai/cuml/issues/1834 so that this refactor can be done in a different PR

JohnZed commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -232,7 +236,48 @@ def predict(self, X, delayed=True, parallelism=5):
+
+        sizes = self.client_.compute([x[1] for x in futures], sync=True)
+
+        models = self.client_.scatter(self.local_model,
```
Do we need this scatter in advance? Wouldn't the predict below it automatically scatter the local_model as needed? I guess this is the optimization to reduce the amount of transfer when you have multiple partitions per worker, which is fair...

cjnolet replied:
There's a few different points to consider here, so I'll address them separately:

1.  Calling `client.submit(self.local_model)` will cause an implicit transfer of the model to the workers and this is okay for smaller models, but specifically the number of parameters for Naive Bayes are on the order  of `n_terms * n_classes`. I didn't find a whole lot of performance difference between doing this and using client.submit, but the warning that's printed is undesirable.   

2. In the completely lazy path, a scatter is done implicitly in the task graph. I have opened https://github.com/rapidsai/cuml/issues/1834 to refactor Naive Bayes like the other algorithms- to extend cuml.Base, use the lazy prediction mixing, etc... The main optimization here is that rather than scattering the objects separately as I was doing before, they can be scattered in one call. That was an oversight on my part.

3. An example of the implicit scatter in the fully-lazy path is here: https://github.com/rapidsai/cuml/blob/branch-0.13/python/cuml/dask/common/base.py#L89. Since `model` is a delayed object, it will be implicitly scattered out to the workers and Dask guarantees only one instance will be made per worker.

cjnolet replied:
Also to add to this- I believe one of the reasons the use of scatter here didn't demonstrate much of a performance difference is because the bulk of the data is stored in GPU and so there really isn't much difference in the serialization / transfer underneath.

cjnolet commented:
@JohnZed @dantegd, I added the conda updates to this PR, if you wouldn't mind giving them a quick look.

jakirkham commented on python/cuml/comm/serialize.py:
```diff
@@ -25,11 +27,37 @@
+
+    # Registers 'dask' serialization by default
+    register_generic(MultinomialNB)
+    register_generic(cuml.Base)
```
Since we are jumping to Dask + Distributed 2.12.0, this can now be simplified.

```suggestion
    register_generic(MultinomialNB, 'cuda',
                     cuda_serialize, cuda_deserialize)
    register_generic(MultinomialNB, 'dask',
                     dask_serialize, dask_deserialize)

    register_generic(cuml.Base, 'cuda',
                     cuda_serialize, cuda_deserialize)
    register_generic(cuml.Base, 'dask',
                     dask_serialize, dask_deserialize)
```

cjnolet replied:
Does `register_generic` still register 'dask' serializer when the serializers aren't provided? If that's the case, can't we simplify this even more?

jakirkham replied:
Yep, sure.

Though the question came up before about whether we want `register_generic(cls)` to have different behavior (like also including `"cuda"`). So this would avoid making any assumptions about what that does.

jakirkham commented on python/cuml/dask/naive_bayes/naive_bayes.py:
```diff
@@ -29,12 +31,8 @@
-
 
-class MultinomialNB(DelayedPredictionMixin):
+class MultinomialNB(object):
```
FWIW in Python 3 all classes inherit from `object`. So this isn't strictly required. Though it's fine to keep this if you wish.

cjnolet commented:
@raydouglass, As mentioned above, I bumped the version of Dask and Distributed to 2.12 so that we can guarantee their new 'cuda' serialization strategies are available.

JohnZed commented on ci/mg/build.sh:
```diff
@@ -51,8 +51,8 @@ conda install -c conda-forge -c rapidsai -c rapidsai-nightly -c nvidia \
       nccl>=2.4 \
-      dask=2.3.0 \
-      distributed=2.3.0 \
+      dask=2.12* \
```
WOW. Glad this got revisited even if it's not actively important today to have the right dask here.

jakirkham replied:
Probably not critical, but would use `dask=2.12.*`. Though it's unlikely we will get to `2.120` without making some kind of breaking change first. 😉

mike-wendt replied:
@cjnolet @jakirkham please match version pinning that cudf uses here and elsewhere 

https://github.com/rapidsai/cudf/pull/4412/files#diff-f60d66f138bb22df65514ede91c1a51dR37

I'm not a fan of conflicting pinning schemes

cjnolet commented:
Unfortunately, some of the recent consolidation of the serialization changes broke linear regression, which is why CI is failing. Once this is fixed, the PR will be ready for merge.

cjnolet commented:
@raydouglass @dillon-cullinan any chance we could get ops approval on the bumping of the Dask versions here?

jakirkham commented:
(asked on Slack)

cjnolet commented:
@raydouglass @jakirkham, interestingly, it looks like `dask_cudf` is still marked `>=2.1.0` for Dask & distributed (https://github.com/rapidsai/cudf/blob/branch-0.13/conda/recipes/dask-cudf/meta.yaml). Is the plan to bump these before code freeze?

kkraus14 commented:
> In the above thread, @kkraus14 points out that `cudf` is going to be pinning to the most recent Dask (>=2.12.x). Is the plan to bump these before code freeze?

Yes that is the plan.

cjnolet commented:
@raydouglass, this CI has been running for several hours now and the TPCxBB team is waiting on the changes. 

In the meantime, I can file a separate PR (and perhaps skip-ci) for adding the most recent Dask/distributed to the `meta.yml` in `conda/recipes`. Are you okay approving this PR as-is?

jakirkham commented:
Where are the Dask + Distributed dependencies expressed in the `cuml` Conda package?

cjnolet commented:
@jakirkham in `conda/recipes`

jakirkham commented:
Right. I'm looking [here]( https://github.com/rapidsai/cuml/blob/branch-0.13/conda/recipes/cuml/meta.yaml ), but don't see `dask` listed. Is that expected or am I looking in the wrong place?

cjnolet commented:
You are looking in the right place. IIRC, we didn't initially pin to a version of Dask, but relied transitively on it through dask-cudf. Originally, dask-cuml was a separate project but now it's merged with cuml and represents a significant amount of our new functionality.

IMO, we probably want to specify versions >=2.12 for Dask and distributed, considering dask-cudf does this. I'd also like to hear thoughts from @JohnZed and @dantegd on this as well.

jakirkham commented:
FWIW we are bumping the requirement in `dask-cudf` ( https://github.com/rapidsai/cudf/pull/4412 ). Not sure if that will be helpful here. Figured it was worth mentioning though.

dantegd commented:
rerun tests

dantegd commented:
rerun tests

jakirkham commented on conda/recipes/cuml/meta.yaml:
```diff
@@ -38,7 +38,7 @@ requirements:
     - python x.x
     - cudf {{ minor_version }}
-    - dask_cudf {{ minor_version }}
+    - dask-cudf {{ minor_version }}
```
Sorry if I'm just misunderstanding. Is this suppose to be `dask-cudf` or `dask-cuda`? Wasn't sure based on the commit message.

dantegd replied:
I don't think us (or cuDF for that matter) have ever explicitly had dask-cuda as a dependency in the recipes, independent of this PR, do you think we should add it?

cjnolet replied:
It probably wouldn't hurt, but I'm sure you both know my suggestion would be for us to do it in a separate PR.

jakirkham replied:
Completely agree. Wasn't trying to hold this up. Just wanted to make sure requiring `dask-cudf` here was intentional 🙂

cjnolet commented:
rerun tests
